doc/index.rst

```diff
@@ -15,6 +15,34 @@ implementations for:
 :mod:`arraycontext` started life as an array abstraction for use with the
 :mod:`meshmode` unstrucuted discretization package.
 
+Design Guidelines
+-----------------
+
+- The array context is about exposing the common subset of operations
+  available in immutable and mutable arrays. As a result, the interface
+  does *not* seek to support interfaces that provide, enable or are typically
+  used only with in-place mutation.
+
+  For example: The equivalents of :func:`numpy.empty` were deprecated
+  and will eventually be removed.
+
+- Each array context offers a specific subset of of numpy under
+  :attr:`arraycontext.ArrayContext.np`. Functions under this namespace
+  must be unconditionally :mod:`numpy`-compatible, that is, they may not
+  offer an interface beyond what numpy offers. Functions that are
+  incompatible, for example by supporting tag metadata
+  (cf. :meth:`arraycontext.ArrayContext.einsum`) should live under the
+  :class:`~arraycontext.ArrayContext` directly.
```

Comment:
Why is this the case? The default argument handling in `actx.einsum` make it compatible with `numpy.einsum`.
(I know there was some discussion about this during early design, but I wasn't heavily involved then. Sorry for the redundancy)

Reply:
I tried to make this explicit. [einsum](https://documen.tician.de/arraycontext/array_context.html#arraycontext.ArrayContext.einsum) does have a `tagged` kwargs, which, by the letter of this rule, disqualifies it from being under `actx.np` (IMO).

Just trying to explain FWIW, the strictness of numpy compatibility is up for debate. Though IMO, it's a slippery slope if we do allow deviations.

Reply:
I see. Yep, that makes sense. Thanks for explaining.